src/parser/parser.ts

```diff
@@ -1,8 +1,242 @@
-import { File } from "../ast/ast";
+import { DatapackDecl, Define, Expression, File, For, Id, If, Import, Let, Print } from "../ast/ast";
 import Options from "../options";
+import Token, { TokenType } from "../ast/token";
+import Lexer from "./lexer";
 
-export default function parse(filename: string, options: Options): File {
-  console.log(filename);
-  console.log(options);
-  throw new Error("Not yet implemented!"); // TODO
+export default class Parser {
+  private filename: string;
+  private lexer: Lexer;
+  options: Options;
+
+  constructor(filename: string, options: Options) {
+    this.filename = filename;
+    this.options = options;
+    this.lexer = new Lexer(filename);
+  }
+  private parseDatapackDeclaration (): DatapackDecl {
+    const datapackToken: Token = this.lexer.lexRegular();
+    this.expect(datapackToken, TokenType.LITERAL, "datapack");
+
+    const id = this.parseId();
+
+    return new DatapackDecl(datapackToken, id);
+  }
+
+  private parseId(): Id {
+    const token: Token = this.lexer.lexRegular();
+    this.expect(token, TokenType.ID);
+
+    return new Id(token);
+  }
+
+  private parseImport() : Import  {
+    const keyword  = this.lexer.lexRegular();
+    // has to be an import
+    this.expect(keyword, TokenType.LITERAL, "import");
+
+    const target = this.parseExpression();
+
+    return new Import(keyword, target);
+  }
+
+  private parseLogicalExpression(): Expression {
+    throw new Error("not implemented yet");
+  }
+
+  private parseDefine(): Expression {
+    const keyword = this.lexer.lexRegular();
+    let id: Id | null = null;
+    const args: Id[] = [];
+
+    this.expect(keyword, TokenType.LITERAL, "import");
+
+    const identifierToken = this.lexer.lexRegular();
+    this.lexer.unlex(identifierToken);
+
+    if (identifierToken.type === TokenType.ID) {
+      id = this.parseId();
+      this.lexer.unlex(identifierToken);
+    }
+
+    const openParan = this.lexer.lexRegular();
+
+    this.expect(openParan, TokenType.LITERAL, "(");
+
+    const arg1 = this.parseId();
+    args.push(arg1);
+
+    let current = this.lexer.lexRegular();
+
+    while (current.content !== ")") {
+      const { type, content, span } = current;
+      const { start } = span;
+      const { line, character } = start;
+
+      switch (type) {
+        case TokenType.ID:
+          this.lexer.unlex(current);
+          const id = this.parseId();
+          args.push(id);
+          break;
+        case TokenType.LITERAL:
+          this.expect(current, TokenType.LITERAL, ",");
+          break;
+        default:
+          throw new ParserError(`${this.filename}: ${line}:${character}: unexpected token '${content}`);
+          break;
```

Comment:
This probably needs to be refactored - you allow multiple commas in a row

Reply:
Right, thanks. I thought about that then completely forgot about it while coding it up.